django/db/models/base.py

```diff
@@ -362,9 +362,18 @@ def _prepare(cls):
         # Set the name of _meta.indexes. This can't be done in
         # Options.contribute_to_class() because fields haven't been added to
         # the model at that point.
+        #
+        # if the name is set with dict argument,
+        # name will be set by format_name_with_model for indexes and constraints.
         for index in cls._meta.indexes:
             if not index.name:
                 index.set_name_with_model(cls)
+            else:
+                index.format_name_with_model(cls)
+
+        for constraint in cls._meta.constraints:
+            if constraint.name:
+                constraint.format_name_with_model(cls)
```

Comment:
I think we should move this to the `Options.contribute_to_class()`. Moreover without cloning constraints and indexes it will not work properly, see falling test
```diff
diff --git a/tests/check_framework/test_model_checks.py b/tests/check_framework/test_model_checks.py
index 0cbc0aff44..9f558f600a 100644
--- a/tests/check_framework/test_model_checks.py
+++ b/tests/check_framework/test_model_checks.py
@@ -236,3 +236,20 @@ class ConstraintNameTests(TestCase):
                 id='models.E032',
             ),
         ])
+
+    @modify_settings(INSTALLED_APPS={'append': 'basic'})
+    @isolate_apps('basic', 'check_framework', kwarg_name='apps')
+    def test_collision_across_apps_special(self, apps):
+        constraint = models.CheckConstraint(check=models.Q(id__gt=0), name='%(app_label)s_%(class)s_foo')
+
+        class Model1(models.Model):
+            class Meta:
+                app_label = 'basic'
+                constraints = [constraint]
+
+        class Model2(models.Model):
+            class Meta:
+                app_label = 'check_framework'
+                constraints = [constraint]
+
+        self.assertEqual(checks.run_checks(app_configs=apps.get_app_configs()), [])
```